doc: design nfs provisioner controlled by operator (#5830)
rook/rook: design/nfs/nfs-provisioner-controlled-by-operator.md

# NFS Provisioner Controlled by Operator

## Summary

> NFS Provisioner is a built in dynamic provisioner for Rook NFS. The functionality works fine but has a issue which the provisioner uses the underlying directory for each provisioned PV when provisioning two or more PV in the same share/export. This issue makes across provisioned PVs in the same share/export has the same data.

Reviewer: ```suggestion
NFS Provisioner is a built in dynamic provisioner for Rook NFS. The functionality works fine but has an issue where the provisioner uses the same underlying directory for each provisioned PV when provisioning two or more PV in the same share/export. This overlap means that each provisioned PV for a share/export can read/write each others data.
```

This hierarchy is the current behaviour of NFS Provisioner when provisioning two PV in the same share/export.

```console
export
├── sample-export
|   ├── data (from PV-A)
|   ├── data (from PV-B)
|   ├── data (from PV-A)
|   └── data (from PV-A)
└── another-export
```

Both PV-A and PV-B uses the `sample-export` directory as their data location.

> This proposal is to make Rook NFS Provisioner to create sub-directory for each provisioned PV when provisioning two or more PV in the same share/export. So it will has a hierarchy like.

Reviewer: ```suggestion
This proposal is to make Rook NFS Provisioner create a sub-directory for each provisioned PV when provisioning two or more PV in the same share/export. So it will have a hierarchy like:
```

> This proposal is to make Rook NFS Provisioner to create sub-directory for each provisioned PV when provisioning two or more PV in the same share/export. So it will has a hierarchy like.

Reviewer: it says we'll create a sub-directory when provisioning 2 or more PV in the same share/export.  Is that implying that we will not create the sub-directory when there's only 1 PV? If so, when the 2nd one is created, do we go back and create a sub-directory for the first PV?

I'm guessing we actually create the sub-directory for every single PV, even when there is only 1. Maybe we should clarify that here if it's truel.

Author: We create sub-directory for every PV. I think i need to edit this sentence. thanks!

```console
export
├── sample-export
│   ├── pv-a
│   │   ├── data (from PV-A)
│   │   ├── data (from PV-A)
│   │   └── data (from PV-A)
│   └── pv-b
│       └── data (from PV-B)
└── another-export
```

> Since those directories are not in the NFS Provisioner container but in NFS Server container, NFS Provisioner cannot directly create sub-directory for that. The solution is by mounting the whole of underlying NFS share/export directory so that the NFS Provisioner can create sub-directory for each provisioned PVs.

Reviewer: ```suggestion
Since those directories are not in the NFS Provisioner pod but in the NFS Server pod, NFS Provisioner cannot directly create sub-directories for them. The solution is to mount the whole underlying NFS share/export directory so that the NFS Provisioner can create a sub-directory for each provisioned PV.
```

### Original Issue

- https://github.com/rook/rook/issues/4982

### Goals

- Make NFS Provisioner to create sub-directory for each provisioned PVs.

> - Make NFS Provisioner to use sub-directory for each provisioned PVs instead of use underlying directory.

Reviewer: ```suggestion
- Make NFS Provisioner use the sub-directory for each provisioned PV instead of using underlying directory.
```

- Improve reliability of NFS Provisioner.

### Non-Goals

- NFS Operator manipulates uncontrolled resources.

## Proposal details

> The approach will be similar to [Kubernetes NFS Client Provisioner](https://github.com/kubernetes-incubator/external-storage/tree/master/nfs-client) which the provisioner mounts the whole of NFS share/export into the provisioner pod (by kubelet) then the provisioner can create the appropriate sub-directory for each provisioned PVs. Currently Rook NFS Provisioner deployed independently and before NFS Server itself we cannot mount the NFS share because we don't know what NFS Server IP and the share/export directory. 

Reviewer: ```suggestion
The approach will be similar to [Kubernetes NFS Client Provisioner](https://github.com/kubernetes-incubator/external-storage/tree/master/nfs-client), where the provisioner mounts the whole of NFS share/export into the provisioner pod (by kubelet), so that the provisioner can then create the appropriate sub-directory for each provisioned PV. Currently Rook NFS Provisioner is deployed independently and before the NFS Server itself, so we cannot mount the NFS share because we don't know the NFS Server IP or the share/export directory. 
```

> The idea is to make NFS Provisioner controlled by the operator. So when NFS Server is created, the operator also creates its provisioner which mount each NFS share/export. And the rest NFS Provisioner can create sub-directory for provisioned PV.

Reviewer: ```suggestion
The idea is to make NFS Provisioner controlled by the operator. So when an NFS Server is created, the operator also then creates its provisioner, which mounts each NFS share/export. Then, the NFS Provisioner can create a sub-directory for each provisioned PV.
```

This is the example NFS Server

```yaml
apiVersion: nfs.rook.io/v1alpha1
kind: NFSServer
metadata:
  name: rook-nfs
  namespace: rook-nfs
spec:
  replicas: 1
  exports:
  - name: share1
    ...
    persistentVolumeClaim:
      claimName: nfs-default-claim
  - name: share2
    ...
    persistentVolumeClaim:
      claimName: nfs-another-claim
```

And the operator will creates the provisioner deployment like

```yaml
kind: Deployment
apiVersion: apps/v1
metadata:
  name: rook-nfs-provisioner
  namespace: rook-nfs
spec:
  ...
    spec:
      ....
      containers:
      - name: rook-nfs-provisioner
        image: rook/nfs:master
        args: ["nfs", "provisioner","--provisioner=nfs.rook.io/nfs-server-provisioner"]
      volumes:
        - name: share1
          nfs:
            server: <NFS_SERVER_IP>
            path: /export/nfs-default-claim
        - name: share2
          nfs:
            server: <NFS_SERVER_IP>
            path: /export/nfs-another-claim
```

> The provisioner deployment will be created on the same namespace with the NFS server and possibility has a same privileges. Since provisioner automatically created by operator the provisioner deployment name and provisioner name flag (`--provisioner`) value will depend on NFSServer name wich deployment name will be added suffix `-provisioner` and provisioner name will started by `nfs.rook.io/` and ended by `-provisioner`

Reviewer: ```suggestion
The provisioner deployment will be created in the same namespace as the NFS server and with the same privileges. Since the provisioner is automatically created by the operator, the provisioner deployment name and provisioner name flag (`--provisioner`) value will depend on NFSServer name. The provisioner deployment name will have an added suffix of `-provisioner` and the provisioner name will start with `nfs.rook.io/`.
```

## Alternatives

> The other possible approach is NFS Provisioner mounts the NFS Server share manually (by executing `mount` command) before creating an appropriate directory for each PV. But in my humble opinion, NFS Provisioner will be lack reliability when met several conditions like NFSServer got exports update, cluster has two or more NFSServer etc.

Reviewer: ```suggestion
The other possible approach is NFS Provisioner mounts the NFS Server share manually (by executing `mount` command) before creating an appropriate directory for each PV. But in my humble opinion, NFS Provisioner would be lacking reliability under several conditions like NFSServer getting its exports updated, the cluster has two or more NFSServer, etc.
```

---
title: nfs-provisioner-controlled-by-operator
target-version: release-X.X
---

# NFS Provisioner Controlled by Operator

## Summary

> Rook NFS Provisioner is a built in dynamic provisioner for Rook NFS. The functionality works fine but has a bug which the provisioner uses the same directory between the provisioned PV.

Reviewer: I think i'm missing something fundamental for this issue, and the mapping between NFSServer, shares/exports, and PVs.  That mapping (it doesn't have to be visual, it could just be textual) could perhaps make this issue and design more clear.

If a user creates a PVC to get a PV that is backed by a NFS export (share)...why is it a concern that another PVC/PV pointing to that same export is backed by the same directory in the provisioner?  NFS shares are read write many, so wouldn't you want them to all be sharing the same underlying data?

The mapping between all these concepts could help clear this up, or please add more background/info so this issue gets more clear about what the problem is and how this design fixes it. Thanks!

Author: > If a user creates a PVC to get a PV that is backed by a NFS export (share)...why is it a concern that another PVC/PV pointing to that same export is backed by the same directory in the provisioner? NFS shares are read write many, so wouldn't you want them to all be sharing the same underlying data?

I think every PV (that is backed by NFS export and provisioned by our provisioner) should have its own data. The problem is when the user creates PV (through rook nfs dynamic provisioner) called PV-A and PV-B. Both of them use the same directory in the `nfs-server export`. Both PV-A and PV-B will have the same data. I mean if the user wants to share data, the user can just use that PVC instead of creating another PVC/PV.

In another case, we will add a deletion feature in our provisioner (which currently not implemented yet). The user creates two PV (from NFS export) called PV-A and PV-B. If both PV-A and PV-B use the same directory for their data and the user deletes the PV-A then what will happen with PV-B? Also about `disk-quota` how provisioner can provide specific storage capacity for each PV while its in one directory?

Reviewer: I see, I was thinking of each user requested PV to map to the root of the share/export that is backing it.  But we're suggesting in this design that instead, each user requested PV should be an independent sub-directory within the share/export.  They are both backed by the same share/export, but they are independent sub-directories that do not overlap.  

Is this hierarchy correct?

* underlying PV (could be Ceph, HostPath, etc.)
  * NFS share/export
    * PV-A provisioned for user 1
    * PV-B provisioned for user 2

In that hierarchy, PV-A and PV-B are backed by the same NFS share/export, but they are independent of each other and do not have any overlap or shared data. Is that correct? 

Please @ mention me when you're updated this design doc. A visual to make these mappings/hierarchy clear would probably be very helpful and effective.

Author: > In that hierarchy, PV-A and PV-B are backed by the same NFS share/export, but they are independent of each other and do not have any overlap or shared data. Is that correct?

Yup, that's correct.

Step to reproduce can be found in this issue [#4982](https://github.com/rook/rook/issues/4982)

### Goals

- Make NFS Provisioner to create different directory for each PV.
- Each NFS Server has own provisioner.
- Improve reliability of NFS Provisioner.

### Non-Goals

- NFS Operator manipulates uncontrolled resources.

## Proposal details

> The approach will be similar to [Kubernetes NFS Client Provisioner](https://github.com/kubernetes-incubator/external-storage/tree/master/nfs-client) which the provisioner mounts the whole of NFS share into the provisioner pod (by kubelet) then the provisioner creates the appropriate directory. Currently Rook NFS Provisioner deployed independently and before NFS Server itself we cannot mount the NFS share because we don't know what NFS Server IP and the export directory.

Reviewer 2: Why do we mount nfs share on the provisioner pod?

Author: So that the provisioner can create the directory for each NFS PV.

Reviewer 2: Okay

Reviewer: Maybe part of my confusion is that *perhaps* we are using language and descriptions that conflates the PVs that back the NFS server shares, and the PVs that a user gets from making a PVC that points (via a storageClass) to a NFS share?

> This proposal is to make NFS Provisioner controlled by the operator. So when NFS Server created, the operator also creates its provisioner which mount each NFS share.

Reviewer 3: ```suggestion
This proposal is to make NFS Provisioner controlled by the operator. So when NFS Server is created, the operator also creates its provisioner which mount each NFS share.
```
